client.go

```diff
@@ -219,122 +286,110 @@ func NewNativeConfig(user, clientVersion string, auth *Auth, hostKeyCallback ssh
+			}
+			sessionInfo.saveConn(conn)
+		} else {
+			// ssh.Client dial does not use a timeout.  In order to make subsequent hops time out, use a separate timer
+			ch := make(chan string, 1)
+			go func() {
+				conn, err = sshClient.Dial("tcp", destAddr)
+				if err != nil {
+					ch <- fmt.Sprintf("ssh client dial fail to %s - %v", destAddr, err)
+				}
+				ch <- ""
+			}()
+			select {
+			case result := <-ch:
+				if result != "" {
+					conn.Close()
+					sessionInfo.CloseAll()
+					return nil, nil, fmt.Errorf(result)
+				}
+			case <-time.After(h.ClientConfig.Timeout):
+				conn.Close()
+				sessionInfo.CloseAll()
+				return nil, nil, fmt.Errorf("ssh client timeout to %s", destAddr)
+			}
+			sshconn, chans, reqs, err := ssh.NewClientConn(conn, h.HostName, h.ClientConfig)
+			if err != nil {
+				return nil, &sessionInfo, fmt.Errorf("NewClientConn fail to %s - %v", destAddr, err)
+			}
+			sshClient = ssh.NewClient(sshconn, chans, reqs)
+			sessionInfo.saveClient(sshClient)
+			sessionInfo.saveConn(conn)
 		}
-		conn = ssh.NewClient(ncc, chans, reqs)
+	} //for
 
-	} else {
-		conn, err = ssh.Dial("tcp", fmt.Sprintf("%s:%d", client.Hostname, client.Port), &client.HostConfig)
-	}
-	if err != nil {
-		return nil, nil, fmt.Errorf("Mysterious error dialing TCP for SSH (we already succeeded at least once) : %s", err)
-	}
-
-	return proxyClient, conn, nil
+	return sshClient, &sessionInfo, nil
 }
 
-func (client *NativeClient) Session() (*ssh.Session, *ssh.Client, *ssh.Client, error) {
-	proxy, conn, err := client.Connect()
+func (nc *NativeClient) Session() (*ssh.Session, *ssh.Client, *SessionInfo, error) {
+	client, sessionInfo, err := nc.Connect()
 	if err != nil {
 		return nil, nil, nil, err
 	}
-	session, err := conn.NewSession()
+	session, err := client.NewSession()
 	if err != nil {
-		conn.Close()
-		if proxy != nil {
-			proxy.Close()
-		}
+		client.Close()
+		sessionInfo.CloseAll()
 		return nil, nil, nil, err
 	}
-	return session, proxy, conn, nil
+	return session, client, sessionInfo, nil
 }
 
 // Output returns the output of the command run on the remote host.
 func (client *NativeClient) Output(command string) (string, error) {
-	session, proxy, conn, err := client.Session()
+	session, conn, sessionInfo, err := client.Session()
+	// even on failure, intermediate hop connections must close
 	if err != nil {
 		return "", err
 	}
+	defer sessionInfo.CloseAll()
 	defer session.Close()
 	defer conn.Close()
-	if proxy != nil {
-		defer proxy.Close()
-	}
-
 	output, err := session.CombinedOutput(command)
-
 	return string(bytes.TrimSpace(output)), wrapError(err)
 }
 
 // Output returns the output of the command run on the remote host as well as a pty.
 func (client *NativeClient) OutputWithPty(command string) (string, error) {
-	session, proxy, conn, err := client.Session()
+	session, conn, sessionInfo, err := client.Session()
 	if err != nil {
 		return "", nil
 	}
+	defer sessionInfo.CloseAll()
```

Comment:
it looks like conn (which is really sshClient, maybe change the name?) gets closed twice, once in sessionInfo.CloseAll() and once on line 392 below. Is that ok? Maybe conn.Close() isn't needed when doing sessionInfo.CloseAll()?

Reply:
it ok to close more than once, but I will review this

Reply:
I think you're right, this stuff is so confusing with the clients, conns, and sessions I was being paranoid.  I remove and do some testing to make sure we don't leak. may not need to return this client at all